contracts/interchain-token/src/contract.rs

```diff
@@ -1,32 +1,365 @@
+            ContractError::InvalidTokenSymbol
+        );
+        Ok(())
+    }
+
+    fn extend_instance_ttl(env: &Env) {
+        env.storage()
+            .instance()
+            .extend_ttl(INSTANCE_TTL_THRESHOLD, INSTANCE_TTL_EXTEND_TO);
+    }
+
+    fn extend_balance_ttl(env: &Env, key: &DataKey) {
+        env.storage()
+            .persistent()
+            .extend_ttl(key, INSTANCE_TTL_THRESHOLD, INSTANCE_TTL_EXTEND_TO);
+    }
+
+    fn read_allowance(env: &Env, from: Address, spender: Address) -> AllowanceValue {
+        let key = DataKey::Allowance(AllowanceDataKey { from, spender });
+        env.storage()
+            .temporary()
+            .get::<_, AllowanceValue>(&key)
+            .map_or(
+                AllowanceValue {
+                    amount: 0,
+                    expiration_ledger: 0,
+                },
+                |allowance| {
+                    if allowance.expiration_ledger < env.ledger().sequence() {
+                        AllowanceValue {
+                            amount: 0,
+                            expiration_ledger: allowance.expiration_ledger,
+                        }
+                    } else {
+                        allowance
+                    }
+                },
+            )
+    }
+
+    fn write_allowance(
+        env: &Env,
+        from: Address,
+        spender: Address,
+        amount: i128,
+        expiration_ledger: u32,
+    ) {
+        let allowance = AllowanceValue {
+            amount,
+            expiration_ledger,
+        };
+
+        assert_with_error!(
+            !(amount > 0 && expiration_ledger < env.ledger().sequence()),
+            "expiration_ledger is less than ledger seq when amount > 0"
+        );
+
+        let key = DataKey::Allowance(AllowanceDataKey { from, spender });
+        env.storage().temporary().set(&key, &allowance);
+
+        if amount > 0 {
+            let live_for = expiration_ledger
+                .checked_sub(env.ledger().sequence())
+                .unwrap();
+
+            env.storage()
+                .temporary()
+                .extend_ttl(&key, live_for, live_for)
+        }
+    }
+
+    fn spend_allowance(env: &Env, from: Address, spender: Address, amount: i128) {
+        let allowance = Self::read_allowance(env, from.clone(), spender.clone());
+
+        assert_with_error!(allowance.amount >= amount, "insufficient allowance");
+
+        if amount > 0 {
+            Self::write_allowance(
+                env,
+                from,
+                spender,
+                allowance
+                    .amount
+                    .checked_sub(amount)
+                    .expect("insufficient allowance"),
+                allowance.expiration_ledger,
+            );
+        }
+    }
+
+    fn read_balance(env: &Env, addr: Address) -> i128 {
+        let key = DataKey::Balance(addr);
+        env.storage()
+            .persistent()
+            .get::<DataKey, i128>(&key)
```

Comment:
```suggestion
            .get::<_, i128>(&key)
```
also, just `get(&key)` doesn't work?